send.go

```diff
@@ -111,8 +111,11 @@ func (s *sender) run(ctx context.Context) error {
 
 func (s *sender) updateProgress(size int, last bool) {
 	if s.progressCb != nil {
+		s.mu.Lock()
 		s.progressCurrent += size
-		s.progressCb(s.progressCurrent, last)
+		progressCurrent := s.progressCurrent // copy to local scope
+		s.mu.Unlock()
+		s.progressCb(progressCurrent, last)
```

Comment:
Technically I think `progressCurrent` could call `progressCb` with a higher value followed by a lower value so keeping a lock longer might make sense.

I'm also not sure if maybe this should be in a goroutine.

Reply:
good point regarding the race.

> I'm also not sure if maybe this should be in a goroutine.

are you thinking something along the lines of `go s.progressCb(progressCurrent, last)` ?

that would introduce a similar race where the receiver isn't guaranteed to get incrementing progress values.